SVG Image art work might not be ready when given to MediaSession

annevk opened the pull request:
#### eb5cdf97ac212fe34e17ae8aa5d5e5f015cdabbe
<pre>
SVG Image art work might not be ready when given to MediaSession
<a href="https://bugs.webkit.org/show_bug.cgi?id=290615">https://bugs.webkit.org/show_bug.cgi?id=290615</a>
<a href="https://rdar.apple.com/148089535">rdar://148089535</a>

Reviewed by Youenn Fablet.

* LayoutTests/fast/mediasession/metadata/artworkdownload-svg-expected.html: Added.
* LayoutTests/fast/mediasession/metadata/artworkdownload-svg.html: Added.

Make sure SVG image has the correct pixels. Use pixelated rendering for
iOS to account for a slight difference in rendering.

* LayoutTests/fast/mediasession/metadata/artworkdownload.html:

Modify existing test to account for it now getting a VideoFrame.

* Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
(WebCore::ArtworkImageLoader::notifyFinished):

Fix the actual bug by using subresourcesAreFinished().

* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.idl:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::loadArtworkImage):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Update test infrastructure so the fix can be tested.

Canonical link: <a href="https://commits.webkit.org/294334@main">https://commits.webkit.org/294334@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97357ab3067c4997ee8d5beae5c7fd6d3fe829ba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21153 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11457 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106641 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52116 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103527 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21460 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29647 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77293 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34319 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104494 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16560 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91650 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57635 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16382 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9669 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51464 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86250 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9743 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108992 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28616 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21050 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86261 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28978 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87857 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85828 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30561 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8271 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22737 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16517 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28547 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33828 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28358 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31678 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29917 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -83,7 +83,12 @@ void ArtworkImageLoader::notifyFinished(CachedResource& resource, const NetworkL
-    m_callback(m_cachedImage->image());
+    Ref image = *m_cachedImage->image();
+    // FIXME: Depending on how https://github.com/w3c/mediasession/issues/346 goes we might want to pass in
+    // (protected) m_document instead of nullptr to subresourcesAreFinished().
```
Passing `nullptr` seems fine here, I don't think we want to delay the onload event.

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -83,7 +83,12 @@ void ArtworkImageLoader::notifyFinished(CachedResource& resource, const NetworkL
+    // FIXME: Depending on how https://github.com/w3c/mediasession/issues/346 goes we might want to pass in
+    // (protected) m_document instead of nullptr to subresourcesAreFinished().
+    image->subresourcesAreFinished(nullptr, [image, callback = std::exchange(m_callback, { })]() mutable {
+        callback(WTFMove(image).ptr());
```
Why do we need WTFMove? Can we write it as `callback(image.ptr())` or do we need `SUPPRESS_UNCOUNTED_ARG`?

youennf commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -7376,13 +7376,17 @@ void Internals::loadArtworkImage(String&& url, ArtworkImagePromise&& promise)
+                else
+                    m_artworkImagePromise->reject(frame.exception().code());
+            } else
+                m_artworkImagePromise->reject(Exception { ExceptionCode::InvalidAccessError, "No native image retrieved."_s });
```
I would tend to write it as:
```
RefPtr document = contextDocument();
if (!document)
    return;
auto promise = std::exchange(m_artworkImagePromise, { });
RefPtr nativeImage = image ? image->nativeImage() : nullptr;
if (!nativeImage) {
    promise->reject(...);
    return;
}
promise->settle(WebCodecsVideoFrame::create(*document, *nativeImage));
```

I would tend to use `weakThis` instead of `protectedThis`.
And maybe we do not need `m_artworkImagePromise` at all and we could just capture it in the `ArtworkImageLoader` callback directly.

shallawa commented on LayoutTests/fast/mediasession/metadata/artworkdownload-svg.html:
```diff
@@ -0,0 +1,22 @@
+<!doctype html>
+<img id=result src="">
```
Try this. It may fix the iOS failure:
```
<img id=result src="" style="image-rendering: pixelated;">
```

annevk committed:
SVG Image art work might not be ready when given to MediaSession
https://bugs.webkit.org/show_bug.cgi?id=290615
rdar://148089535

Reviewed by Youenn Fablet.

* LayoutTests/fast/mediasession/metadata/artworkdownload-svg-expected.html: Added.
* LayoutTests/fast/mediasession/metadata/artworkdownload-svg.html: Added.

Make sure SVG image has the correct pixels. Use pixelated rendering for
iOS to account for a slight difference in rendering.

* LayoutTests/fast/mediasession/metadata/artworkdownload.html:

Modify existing test to account for it now getting a VideoFrame.

* Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
(WebCore::ArtworkImageLoader::notifyFinished):

Fix the actual bug by using subresourcesAreFinished().

* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.idl:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::loadArtworkImage):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Update test infrastructure so the fix can be tested.

Canonical link: https://commits.webkit.org/294334@main
